theano/gpuarray/tests/test_linalg.py

```diff
@@ -112,3 +112,105 @@ def test_linalgerr_solve(self):
+        cholesky_op = GpuCholesky(lower=lower, inplace=inplace)
+        chol_A = cholesky_op(A)
+        return theano.function([A], chol_A, accept_inplace=inplace, mode=mode_with_gpu)
+
+    def compare_gpu_cholesky_to_numpy(self, A_val, lower=True, inplace=False):
+        """ Helper function to compare op output to numpy.cholesky output. """
+        chol_A_val = numpy.linalg.cholesky(A_val)
+        if not lower:
+            chol_A_val = chol_A_val.T
+        fn = self.get_gpu_cholesky_func(lower, inplace)
+        res = fn(A_val)
+        chol_A_res = numpy.array(res)
+        utt.assert_allclose(chol_A_res, chol_A_val)
+
+    def test_invalid_input_fail_non_square(self):
+        """ Invalid Cholesky input test with non-square matrix as input. """
+        A_val = numpy.random.normal(size=(3, 2)).astype("float32")
+        fn = self.get_gpu_cholesky_func(True, False)
+        self.assertRaises(ValueError, fn, A_val)
+
+    def test_invalid_input_fail_vector(self):
+        """ Invalid Cholesky input test with vector as input. """
+        def invalid_input_func():
+            A = theano.tensor.vector("A", dtype="float32")
+            GpuCholesky(lower=True, inplace=False)(A)
+        self.assertRaises(AssertionError, invalid_input_func)
+
+    def test_invalid_input_fail_tensor3(self):
+        """ Invalid Cholesky input test with 3D tensor as input. """
+        def invalid_input_func():
+            A = theano.tensor.tensor3("A", dtype="float32")
+            GpuCholesky(lower=True, inplace=False)(A)
+        self.assertRaises(AssertionError, invalid_input_func)
+
+    def test_diag_chol(self):
+        """ Diagonal matrix input Cholesky test. """
+        # make sure all diagonal elements are positive so positive-definite
+        A_val = numpy.diag(numpy.random.uniform(size=5).astype("float32") + 1)
+        self.compare_gpu_cholesky_to_numpy(A_val, lower=True, inplace=False)
+
+    def test_dense_chol_lower(self):
+        """ Dense matrix input lower-triangular Cholesky test. """
+        M_val = numpy.random.normal(size=(3, 3)).astype("float32")
+        # A = M.dot(M) will be positive definite for all non-singular M
+        A_val = M_val.dot(M_val.T)
+        self.compare_gpu_cholesky_to_numpy(A_val, lower=True, inplace=False)
+
+    def test_dense_chol_upper(self):
+        """ Dense matrix input upper-triangular Cholesky test. """
+        M_val = numpy.random.normal(size=(3, 3)).astype("float32")
+        # A = M.dot(M) will be positive definite for all non-singular M
+        A_val = M_val.dot(M_val.T)
+        self.compare_gpu_cholesky_to_numpy(A_val, lower=False, inplace=False)
+
+    def test_diag_chol_inplace(self):
+        """ Diagonal matrix input inplace Cholesky test. """
+        # make sure all diagonal elements are positive so positive-definite
+        A_val = numpy.diag(numpy.random.uniform(size=5).astype("float32") + 1)
+        self.compare_gpu_cholesky_to_numpy(A_val, lower=True, inplace=True)
+
```

Comment:
It miss some `test_diag_chol_*()` for `lower=False, inplace=True` and `lower=False, inplace=False` !

Also, consider factorizing the code for valid test cases (diagonal matrices and dense matrices). Maybe you could write one function for each type of matrix (e.g. `run_test_diag(inplace, lower)` and `run_test_dense(inplace, lower)`), and then use it in every combination of lower and inplace.